### PR TITLE
Player Statistics

### PR DESCRIPTION
**Pull request by StarWitch**
Forked off of #194, as per our discussion in Discord.

Not much has changed except for the fact that now I'm using the REPLAY_GFLAG* typedef to save booleans rather than writing a bunch of booleans to the replay file one at a time.

This way we can also have additional game modes, assist mode options, "comedy options" etc. in their own separate framework without being tied together with this.

Has some rough edges but I'm mostly looking for a yay/nay before I clean it up.

### CODE REVIEW

File: src/replay.h [hunk 1]
```diff
@@ -130,6 +133,15 @@ typedef struct ReplayStage {
 	uint64_t plr_points_final;
 	/* END REPLAY_STRUCT_VERSION_TS102000_REV3 and above */
 
+	/* BEGIN REPLAY_STRUCT_VERSION_TS104000_REV0 and above */
+	uint8_t plr_stats_enabled;
```
Comment (Akaricchi) [inline src/replay.h:137]: This one is not used I think?

File: src/stagedraw.c [hunk 2]
```diff
@@ -1826,7 +1826,7 @@ void stage_display_clear_screen(const StageClearBonus *bonus) {
 
 	stagetext_add(
 		"Press Fire to continue",
-		VIEWPORT_W/2 + VIEWPORT_H*0.7*I,
+		VIEWPORT_W/2 + VIEWPORT_H*0.8*I,
```
Comment (Akaricchi) [inline src/stagedraw.c:1829]: Unrelated change?

File: src/stagetext.c [hunk 3]
```diff
@@ -187,16 +187,6 @@ void stagetext_table_add_separator(StageTextTable *tbl) {
...
-void stagetext_table_test(void) {
-	StageTextTable tbl;
-	stagetext_begin_table(&tbl, "Test", RGB(1, 1, 1), RGB(1, 1, 1), VIEWPORT_W/2, 60, 300, 30, 60);
-	stagetext_table_add(&tbl, "foo", "bar");
-	stagetext_table_add(&tbl, "qwerty", "asdfg");
-	stagetext_table_add(&tbl, "top", "kek");
-	stagetext_table_add_separator(&tbl);
-	stagetext_table_add_numeric(&tbl, "Total Score", 9000000);
-	stagetext_end_table(&tbl);
-}
```
Comment (Akaricchi) [inline src/stagetext.c:199]: This is also unrelated, but it's dead code so I don't mind it gone; I guess it's fine.

File: src/stats.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * This software is licensed under the terms of the MIT License.
+ * See COPYING for further information.
+ * ---
+ * Copyright (c) 2011-2019, Lukas Weber <laochailan@web.de>.
+ * Copyright (c) 2012-2019, Andrei Alexeyev <akari@taisei-project.org>.
+ */
+
+#include "taisei.h"
+
+#include "config.h"
```
Comment (Akaricchi) [inline src/stats.c:11]: Not needed

File: src/stats.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "stats.h"
+
+#include "global.h"
+#include "stage.h"
```
Comment (Akaricchi) [inline src/stats.c:16]: Probably also not needed. For the log stuff, just include `log.h`.

File: src/stats.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+#ifndef IGUARD_stats_h
+#define IGUARD_stats_h
+
+#include "taisei.h"
+
+typedef struct Stats Stats;
+
+struct Stats {
+	struct {
+		int lives;
```
Comment (Akaricchi) [inline src/stats.h:18]: I think this is a little misleading, because it can be interpreted as either "lives used" or "total lives gained". I would rename it either `deaths` or `lives_used`; perhaps the same applies to bombs.

File: src/player.c [hunk 7]
```diff
@@ -885,6 +900,7 @@ void player_realdeath(Player *plr) {
 	plr->bomb_fragments = 0;
 	plr->voltage *= 0.9;
 	plr->lives--;
+	stats_append_life(&plr->stats);
```
Comment (Akaricchi) [inline src/player.c:903]: Case in point: I was pretty confused about this code fragment, appending a "life" when a "death" has occurred, until I realized what the correct interpretation is.

File: src/stats.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+		int bombs;
+		int continues;
+	} total, stage;
+};
+
+void stats_init(Stats *stats);
+
+void stats_append_life(Stats *stats);
+void stats_append_bomb(Stats *stats);
+void stats_append_continue(Stats *stats);
```
Comment (Akaricchi) [inline src/stats.h:28]: Also, I would suggest instead of "append" to use the word "track", as in `stats_track_<event>`. `stats_track_life_lost`, `stats_track_bomb_used`, reads more naturally IMO.

File: src/menu/common.c [hunk 9]
```diff
@@ -91,6 +91,7 @@ static void reset_game(StartGameContext *ctx) {
 	replay_destroy(&global.replay);
 	replay_init(&global.replay);
 	player_init(&global.plr);
+	stats_init(&global.plr.stats);
```
Comment (Akaricchi) [inline src/menu/common.c:94]: I think you also need this call in `main_singlestg_begin_game` (`main.c`). That is where game initialization is done when starting a stage from the command line (`taisei --play --sid <stageID>`).

(Ideally we should really improve our terrible game-session abstraction (or lack thereof), but that's out of scope for this PR)

File: src/stats.c [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+	stats->total.continues++;
+	stats->stage.continues++;
+	log_debug("bombs used (total): %d", stats->total.continues);
+	log_debug("bombs used (stage): %d", stats->stage.continues);
+}
+
+void stats_stage_reset(Stats *stats) {
+	stats->stage.lives = 0;
+	stats->stage.bombs = 0;
+	stats->stage.continues = 0;
```
Comment (Akaricchi) [inline src/stats.c:46]:
```suggestion
	memset(&stats->stage, 0, sizeof(stats->stage));
```

File: src/stats.c (hunk 4 above)
Comment (Akaricchi) [reply src/stats.c:11]: Clarification: this comment only applies to the `#include "config.h"` line; `taisei.h` must be included by every file.

**Comment by StarWitch** [comment]
Current round of review done! Some of that was just me being tired and sloppy. Thanks for catching it.
